src/tsa_products/tests.py

```diff
@@ -1,3 +1,335 @@
+            detail_image=self.test_detail_image,
+            is_uploaded=self.test_is_uploaded,
+        )
+        # Runs model validation for given data.
+        product.full_clean()
+        self.assertEqual(Product.objects.count(), 1)
+
+    def test_failure_product_creation_without_category(self):
+        # Test the failure of product creation without a category.
+        with self.assertRaisesMessage(ValidationError, "{'category': ['This field cannot be null.']}"):
+            product = Product(title=self.test_title)
+            product.full_clean()
+            product.save()
+        self.assertEqual(Product.objects.count(), 0)
+
+    def test_failure_product_creation_without_title(self):
+        # Test the failure of product creation without a title.
+        with self.assertRaisesMessage(ValidationError, "{'title': ['This field cannot be blank.']}"):
+            product = Product(category=self.test_category)
+            product.full_clean()
+            product.save()
+        self.assertEqual(Product.objects.count(), 0)
+
+
+class ProductVariationTestCase(TestCase):
+
+    @classmethod
+    def setUpTestData(self):
+        # Creates testing data for the test cases.
+        self.test_product = Product.objects.create(
+            category=Category.objects.create(title="test", image="test-path"), title="test-title"
+        )
+        self.test_product_color = ProductColor.objects.create(color_nickname="test-color-name")
+        self.test_amount = 1
+
+    def setUp(self):
+        # Ensures there are categories in the db before a test case is running.
+        ProductVariation.objects.all().delete()
+
+    def test_successful_product_variation_creation(self):
+        # Tests the creation of a product variation.
+        product_variation = ProductVariation.objects.create(
+            product=self.test_product, product_color=self.test_product_color, amount=self.test_amount
+        )
+        # Runs model validation for given data.
+        product_variation.full_clean()
+        self.assertEqual(ProductVariation.objects.count(), 1)
+
+    def test_failure_product_variation_creation_without_product(self):
+        # Test the failure of product variation creation without a product.
+        with self.assertRaisesMessage(ValidationError, "{'product': ['This field cannot be blank.']}"):
+            product_variation = ProductVariation(amount=self.test_amount)
+            product_variation.full_clean()
+            product_variation.save()
+        self.assertEqual(ProductVariation.objects.count(), 0)
+
+    def test_success_get_all_lettering_items_method(self):
+        test_product_variation = ProductVariation.objects.create(product=self.test_product)
+        item1 = LetteringItemVariation.objects.create(product_variation=test_product_variation)
+        item2 = LetteringItemVariation.objects.create(product_variation=test_product_variation)
+
+        items = test_product_variation.get_all_lettering_items()
```

Comment:
1. would rename to `all_items`
2. (optional, if default sorting should be `by_id`) add a variable `sorted_items` that should be sorted by the IDs of the queryset values (like in the assert below)